File: sample_code.R
# =============================================
# Sample R Code
# Author: https://github.com/xiuren112
# Date: 2026-01-05
# Description: Utility functions with basic error handling and examples
# =============================================

# -----------------------------
# Function 1: Add two numbers
# -----------------------------
add_numbers <- function(a, b) {
  if (!is.numeric(a) || !is.numeric(b)) {
    stop("Both 'a' and 'b' must be numeric")
  }
  return(a + b)
}

# Example usage:
# result <- add_numbers(2, 3)
# print(result)  # 5

# -----------------------------
# Function 2: Count words in a string
# -----------------------------
count_words <- function(text) {
  if (!is.character(text)) {
    stop("'text' must be a character vector")
  }
  # Count non-space sequences
  return(sum(gregexpr("\\S+", text)[[1]] > 0))
}

# Example usage:
# words <- count_words("Hello world from R")
# print(words)  # 4

# -----------------------------
# Function 3: Safe division
# -----------------------------
safe_divide <- function(a, b) {
  if (!is.numeric(a) || !is.numeric(b)) {
    stop("Both 'a' and 'b' must be numeric")
  }
  if (b == 0) {
    warning("Division by zero! Returning NA.")
    return(NA)
  }
  return(a / b)
}

# Example usage:
# print(safe_divide(10, 2))  # 5
# print(safe_divide(10, 0))  # NA with warning

File: sample_code.go
// =============================================
// Sample Go Code
// Author: https://github.com/xiuren112
// Date: 2026-01-05
// Description: Utility functions with basic error handling and examples
// =============================================

package main

import (
	"errors"
	"fmt"
	"strings"
)

// -----------------------------
// Function 1: Add two numbers
// -----------------------------
func AddNumbers(a, b float64) float64 {
	return a + b
}

// -----------------------------
// Function 2: Count words in a string
// -----------------------------
func CountWords(text string) int {
	words := strings.Fields(text)
	return len(words)
}

// -----------------------------
// Function 3: Safe division
// -----------------------------
func SafeDivide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, errors.New("division by zero")
	}
	return a / b, nil
}

// -----------------------------
// Main function with examples
// -----------------------------
func main() {
	// Example 1: AddNumbers
	sum := AddNumbers(5, 7)
	fmt.Println("AddNumbers(5,7) =", sum)

	// Example 2: CountWords
	count := CountWords("Hello world from Go")
	fmt.Println("CountWords('Hello world from Go') =", count)

	// Example 3: SafeDivide
	result, err := SafeDivide(10, 2)
	if err != nil {
		fmt.Println("SafeDivide(10,2) error:", err)
	} else {
		fmt.Println("SafeDivide(10,2) =", result)
	}

	// Example 3b: division by zero
	result, err = SafeDivide(10, 0)
	if err != nil {
		fmt.Println("SafeDivide(10,0) error:", err)
	} else {
		fmt.Println("SafeDivide(10,0) =", result)
	}
}
